src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+                .parse::<usize>()
+                .unwrap()
+        });
+
+        Ok(segments)
+    }
+
+    fn find_unfinished_wal_segment(&self) -> Result<PathBuf, anyhow::Error> {
+        lazy_static! {
+            static ref UNFINISHED_WAL_SEGMENT_REGEX: Regex = Regex::new("^log-[0-9]+$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &UNFINISHED_WAL_SEGMENT_REGEX)?;
+        if segments.len() > 1 {
+            bail!(
+                "Expected only a single unfinished wal segment at {}. Found {}",
+                self.wal_path.display(),
+                segments.len()
+            );
+        }
+
+        if segments.len() == 0 {
+            bail!(
+                "Expected at least a single unfinished wal segment at {}. Found none.",
+                self.wal_path.display()
+            );
+        }
+
+        Ok(segments.pop().unwrap())
+    }
+
+    fn find_batches(&self) -> Result<Vec<Batch>, anyhow::Error> {
+        lazy_static! {
+            static ref BATCH_REGEX: Regex = Regex::new("^batch-[0-9]+-[0-9]+$").unwrap();
+        }
+
+        let batches = read_dir_regex(&self.trace_path, &BATCH_REGEX)?;
+
+        let mut batches: Vec<Batch> = batches
+            .into_iter()
+            .map(Batch::reinit)
+            .collect::<Result<_, _>>()
+            .unwrap();
+        batches.sort_by_key(|batch| batch.lower);
+        Ok(batches)
+    }
+
+    /// Try to compact all of the batches we know about into a single batch from
+    /// [compaction_frontier, upper)
+    ///
+    /// TODO: the approach to compacting is likely very suboptimal.
+    fn compact(&mut self) -> Result<(), anyhow::Error> {
+        self.consume_wal()?;
+
+        if self.batches.len() > 10 {
+            let batches = std::mem::replace(&mut self.batches, vec![]);
+            let batch = Batch::compact(&batches, &self.trace_path, self.compaction)?;
+            self.batches.push(batch);
+
+            // TODO: This seems like potentially a place with a weird failure mode.
+            for batch in batches {
+                fs::remove_file(&batch.path).with_context(|| {
+                    format!("failed to remove replaced batch {}", batch.path.display())
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Re-initialize a trace based on the available batch files on disk.
+    pub fn resume(
+        id: GlobalId,
+        traces_path: &Path,
+        wals_path: &Path,
+    ) -> Result<Self, anyhow::Error> {
+        // Need to instantiate a new trace and figure out what batches
+        // we have access to.
+        let trace_path = traces_path.join(id.to_string());
+        let wal_path = wals_path.join(id.to_string());
+        let mut ret = Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        };
+
+        let mut batches = ret.find_batches()?;
+        ret.batches.append(&mut batches);
```

Comment:
looks like this could have been `ret.batches = ret.find_batches()?` but isn't for some reason? Is there a reason `self.find_batches()` doesn't just hydrate `self.batches`?

Reply:
mainly to make find_batches more parallel with find_finished / unfinished wal_segment. changed logic here to be simpler